python/pyspark/sql/tests/connect/streaming/test_parity_listener.py

```diff
@@ -19,38 +19,153 @@
 import time
 
 from pyspark.sql.tests.streaming.test_streaming_listener import StreamingListenerTestsMixin
-from pyspark.sql.streaming.listener import StreamingQueryListener, QueryStartedEvent
-from pyspark.sql.types import StructType, StructField, StringType
+from pyspark.sql.streaming.listener import (
+    StreamingQueryListener,
+    QueryStartedEvent,
+    QueryProgressEvent,
+    QueryIdleEvent,
+    QueryTerminatedEvent,
+)
+from pyspark.sql.types import (
+    ArrayType,
+    StructType,
+    StructField,
+    StringType,
+    IntegerType,
+    FloatType,
+    MapType,
+)
+from pyspark.sql.functions import count, lit
 from pyspark.testing.connectutils import ReusedConnectTestCase
 
 
 def get_start_event_schema():
     return StructType(
         [
-            StructField("id", StringType(), True),
-            StructField("runId", StringType(), True),
+            StructField("id", StringType(), False),
+            StructField("runId", StringType(), False),
             StructField("name", StringType(), True),
-            StructField("timestamp", StringType(), True),
+            StructField("timestamp", StringType(), False),
         ]
     )
```

Comment:
I'm thinking of just move these methods to the `QueryxxxEvent`, maybe even create a method that called `asDataFrame`. To save user's effort to create this by themselves. For example, before:

```
def get_start_event_schema():
    return StructType(
        [
            StructField("id", StringType(), False),
            StructField("runId", StringType(), False),
            StructField("name", StringType(), True),
            StructField("timestamp", StringType(), False),
        ]
    )

class TestListener(StreamingQueryListener):
    def onQueryStarted(self, event):
        df = self.spark.createDataFrame(
            data=[(event.asDict())],
            schema=get_start_event_schema(),
        )
        df.write.saveAsTable("listener_start_events")
```
Note that this looks simple because I wrote the `asDict` method, and test `get_start_event_schema` method. In production, users need to do this themselves. But these are really redundant, if we manage these:

For example:

```
class TestListener(StreamingQueryListener):
    def onQueryStarted(self, event):
        df = self.spark.createDataFrame(
            data=[(event.asDict())],
            schema=event.schema(),
        )
        df.write.saveAsTable("listener_start_events")

============= OR ===============
class TestListener(StreamingQueryListener):
    def onQueryStarted(self, event):
        df = event.asDataFrame(self.spark)
        df.write.saveAsTable("listener_start_events")
```

Reply:
What do you think? @rangadi  @bogao007

Reply:
Also cc @HyukjinKwon @HeartSaVioR

Reply:
yeah I saw the below code that constructs schemas, it is a pain... I think it would be good if we could simplify this in a certain way, this would save a lot of redundant work on both customer side and our side (writing tests/ internal team usage).

Reply:
I think for this one, we can drop None values in for loop inside `_invoke_function()`. There is no impact on Protobuf. 
But not needed in the PR. Can be done as improvement.

Reply:
@rangadi are you referring to [this](https://github.com/apache/spark/pull/42563) PR? I think you commented in the wrong PR :)

Reply:
Sorry about that. Removed my comment.

Reply:
Yeah we can change like that - I guess that's the only way to fix the test?

Reply:
@HyukjinKwon 
I'm looking at the [test error here](https://github.com/WweiL/oss-spark/actions/runs/5884139868/job/15959887335) -- I couldn't reproduce it locally. 

But I think the change is orthogonal to the test error. It's more about an addition to the listener events API. We can just define the `asDict`, and `get_event_schema` method in the test suite. And the test still runs.

But that would mean users need to add exactly the same redundant code I added in the suite if they want to write the event to external table. For example, before this addition, the `onQueryStartedEvent` is implemented like this:
https://github.com/apache/spark/blob/master/python/pyspark/sql/tests/connect/streaming/test_parity_listener.py#L27-L44

That looks not as painful, but `onQueryProgress` would be extremely painful I think. 

Because it's very likely every user who want to write events to external tables need to redo the same code all times, I'm thinking providing the API so they don't need to reinvent the wheel

Reply:
Synced with @rangadi, he thinks this is not needed. I'll put everything into test suite